File: 96/UniqueBST.java
public class Solution {
	public int numTrees(int n) {
		if (n <= 0) return 0;

		int[] mem = new int[n];
		return helper(n, mem);
	}

	private int helper(int n, int[] mem) {

		if (n == 0)
			return 1;
		if (mem[n-1] != 0) 
			return mem[n-1];

		int sum = 0;
		for (int i=1; i<=n; i++) {
			sum += 
				helper(i-1, mem)
					*
				helper(n-i, mem);
		}

		mem[n-1] = sum;
		return sum;
	}
}

File: 096/unique_bst.py
class Solution:

    def numTrees(self, n: int) -> int:
        cache = {0: 1, 1: 1}
        return self._numTrees(cache, n)

    def _numTrees(self, cache: dict, n: int) -> int:
        if n in cache:
            return cache[n]
        
        ret = 0
        for i in range(0, n):
            n_left = i
            n_right = n-i-1
            left = cache.get(
                n_left,
                self._numTrees(cache, n_left),
            )
            right = cache.get(
                n_right,
                self._numTrees(cache, n_right),
            )
            ret += left * right

        cache[n] = ret

        return ret
